google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/SpannerDialect.java

```diff
@@ -95,170 +95,398 @@ public SpannerDialect() {
-    registerFunction("FLOOR", new StandardSQLFunction("FLOOR", StandardBasicTypes.DOUBLE));
-    registerFunction("COS", new StandardSQLFunction("COS", StandardBasicTypes.DOUBLE));
-    registerFunction("COSH", new StandardSQLFunction("COSH", StandardBasicTypes.DOUBLE));
-    registerFunction("ACOS", new StandardSQLFunction("ACOS", StandardBasicTypes.DOUBLE));
-    registerFunction("ACOSH", new StandardSQLFunction("ACOSH", StandardBasicTypes.DOUBLE));
-    registerFunction("SIN", new StandardSQLFunction("SIN", StandardBasicTypes.DOUBLE));
-    registerFunction("SINH", new StandardSQLFunction("SINH", StandardBasicTypes.DOUBLE));
-    registerFunction("ASIN", new StandardSQLFunction("ASIN", StandardBasicTypes.DOUBLE));
-    registerFunction("ASINH", new StandardSQLFunction("ASINH", StandardBasicTypes.DOUBLE));
-    registerFunction("TAN", new StandardSQLFunction("TAN", StandardBasicTypes.DOUBLE));
-    registerFunction("TANH", new StandardSQLFunction("TANH", StandardBasicTypes.DOUBLE));
-    registerFunction("ATAN", new StandardSQLFunction("ATAN", StandardBasicTypes.DOUBLE));
-    registerFunction("ATANH", new StandardSQLFunction("ATANH", StandardBasicTypes.DOUBLE));
-    registerFunction("ATAN2", new StandardSQLFunction("ATAN2", StandardBasicTypes.DOUBLE));
-  }
-
-  @Override
-  public Exporter<Table> getTableExporter() {
+  @Override
+  public void initializeFunctionRegistry(SqmFunctionRegistry registry) {
+    super.initializeFunctionRegistry(registry);
+
+    // Aggregate Functions
+    registry.namedTemplateBuilder("ANY_VALUE")
+        .setExactArgumentCount(1)
+        .register();
+    registry.namedTemplateBuilder("ARRAY_AGG")
+        .setExactArgumentCount(1)
+        .register();
+    registry.namedTemplateBuilder("BIT_AND")
+        .setInvariantType(StandardSpiBasicTypes.LONG)
+        .setExactArgumentCount(1)
+        .register();
+    registry.namedTemplateBuilder("BIT_OR")
+        .setInvariantType(StandardSpiBasicTypes.LONG)
+        .setExactArgumentCount(1)
+        .register();
+    registry.namedTemplateBuilder("BIT_XOR")
+        .setInvariantType(StandardSpiBasicTypes.LONG)
+        .setExactArgumentCount(1)
+        .register();
+    registry.namedTemplateBuilder("COUNTIF")
+        .setInvariantType(StandardSpiBasicTypes.LONG)
+        .setExactArgumentCount(1)
+        .register();
+    registry.namedTemplateBuilder("LOGICAL_AND")
+        .setInvariantType(StandardSpiBasicTypes.BOOLEAN)
+        .setExactArgumentCount(1)
+        .register();
+    registry.namedTemplateBuilder("LOGICAL_OR")
+        .setInvariantType(StandardSpiBasicTypes.BOOLEAN)
+        .setExactArgumentCount(1)
+        .register();
+    registry.namedTemplateBuilder("STRING_AGG")
+        .setInvariantType(StandardSpiBasicTypes.STRING)
+        .setArgumentCountBetween(1, 2)
+        .register();
+
+    // Mathematical Functions
+    CommonFunctionFactory.abs(registry);
```

Comment:
I don't see this one in the list above that you replaced from ORM 5.
It looks like you referenced the list of functions provided [here](https://cloud.google.com/spanner/docs/functions-and-operators). ORM 5 list doesn't seem to follow the same approach. Can we ensure consistency?

Reply:
@dmitry-s How did you approach the original list in ORM 5 dialect to ensure completeness?

Reply:
I got the list of functions from the documentation.

Reply:
@meltsufin You can't see it here because it is inherited from the parent class.

Reply:
@dzou Did the parent class change for ORM 6?

Reply:
Same parent class, but they have different ways of implementing function mappings.

Hibernate 5 way: https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java#L162

Hibernate 6 way: https://github.com/hibernate/hibernate-orm/blob/wip/6.0/hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java#L283

Pretty sure both dialects are equivalent. I did this port by writing a script and running a regex expression on the Spanner Hibernate 5 mappings and then converted them to Hibernate 6 new style. Then manually verified + compared with [postgres Hibernate 6 dialect](https://github.com/hibernate/hibernate-orm/blob/wip/6.0/hibernate-core/src/main/java/org/hibernate/dialect/PostgreSQL81Dialect.java#L113). That's where the CommonFunctionRegistry stuff was used.

Reply:
cool!